rfcs/remote_channel.md

```diff
@@ -0,0 +1,616 @@
+type. In cases like arrays where there is a `value` field holding a
+container object, `RemoteObject.toLocal()` recursively converts the
+content of the container into local objects (so e.g. a `type: array` is
+converted into a local `Array` instances, and any contained `type: object`
+objects are converted into local `Object` instances, and so on through
+the full tree).
+
+#### Higher Level API
+
+The low-level API provides required primitives, but it's difficult to
+use directly. To achieve the aim of making tests no harder to write
+than SpecialPowers-based Gecko tests, there is also a higher-level API
+initially providing two main capabilities: the ability to
+`postMessage` a remote context, and the ability to `executeScript` so
+that the script runs in a remote context.
+
+This API is provided by a `RemoteWindow` object. The `RemoteWindow`
+object doesn't handle creating the browsing context, but given the
+UUID for the remote window, creates a `SendChannel` which is able to
+send messages to the remote. Alternatively the `RemoteWindow` may be
+created first and its `uuid` property used when constructing the URL.
+
+Inside the remote browsing context itself, the test author has to call
+`await start_window()` in order to set up a `RecvChannel` with UUID given
+by the `uuid` parameter in `location.href`. The returned object offers
+an `addMessageHandler(callback)` API to receive messages sent with the
+`postMessage` API on `RemoteWindow`, and `async nextMessage()` to wait
+for the next message.
+
+##### Script Execution
+
+`RemoteWindow.executeScript(fn, ...args)` serializes the function
+`fn`, using `Function.toString()`. Each argument is
+serialized using the remote value serialization algorithm. Along with
+the function string and the arguments, a `SendChannel` is sent for the
+command response (only one such channel is created per `RemoteWindow`
+for efficiency reasons), and a command id is sent to disambiguate
+responses from different commands.
+
+On the remote side the function is deserialized and executed. If
+execution results in a `Promise` value, the result of that promise is
+awaited. The final return value after the promise is resolved is sent
+back and forms the async return value of the `executeScript` call. If
+the script throws, the thrown value is provided as the result, and
+re-thrown in the originating context. In addition an `exceptionDetails`
+field provides the line/column numbers of the original exception,
+where available.
+
+In addition there is a `RemoteWindow.executeScriptNoResult(fn,
+...args)` method. This works the same way except no channel is passed,
+and so no result is returned. This can be useful in case the script
+does something like trigger a navigation, so there's no need to
+synchronize the navigation starting (which will close the socket) with
+writing the response.
+
+TODO: the naming here isn't great. In particular a `RemoteWindow`
```

Comment:
Doesn't `RemoteContext` work (as proposed in https://github.com/web-platform-tests/rfcs/pull/91)?

Reply:
No, because `RemoteContext` is already used by testharness.js to mean something slightly different:
```
    /*
     * A RemoteContext listens for test events from a remote test context, such
     * as another window or a worker. These events are then used to construct
     * and maintain RemoteTest objects that mirror the tests running in the
     * remote context.
     */
```

In the long term it would make sense to unify these features, but in the short term a naming conflict will cause breakage.